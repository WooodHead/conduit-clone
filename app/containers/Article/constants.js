/*
 *
 * Article constants
 *
 */

export const DEFAULT_ACTION = 'app/Article/DEFAULT_ACTION';
export const FETCH_ARTICLE = 'app/Article/FETCH_ARTICLE';
export const ARTICLE_LOADED = 'app/Article/ARTICLE_LOADED';
export const LOAD_ARTICLE_ERROR = 'app/Article/LOAD_ARTICLE_ERROR';
