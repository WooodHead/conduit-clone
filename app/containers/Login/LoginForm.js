import React from 'react';
import PropTypes from 'prop-types';
import { Form, Field, reduxForm } from 'redux-form/immutable';
import Input from 'components/Input';
import { Button } from 'reactstrap';

const required = (value) => (value ? undefined : 'Required');
const email = (value) =>
  value && !/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i.test(value)
    ? 'Invalid email address'
    : undefined;

let LoginForm = (props) => { // eslint-disable-line import/no-mutable-exports
  const { handleSubmit, pristine, reset, submitting } = props;
  return (
    <Form onSubmit={handleSubmit}>
      <div>
        <label htmlFor="email">Email</label>
        <Field
          name="email"
          component={Input}
          type="text"
          validate={[required, email]}
        />
      </div>
      <div>
        <label htmlFor="password">Password</label>
        <Field
          name="password"
          component={Input}
          type="password"
          validate={[required]}
        />
      </div>
      <Button type="submit" outline color="primary" disabled={submitting}>
        Submit
      </Button>{' '}
      <Button outline color="primary" disabled={pristine || submitting} onClick={reset}>
        Clear Values
      </Button>
    </Form>
  );
};

LoginForm.propTypes = {
  handleSubmit: PropTypes.func,
  reset: PropTypes.func,
  pristine: PropTypes.bool,
  submitting: PropTypes.bool,
};

LoginForm = reduxForm({
  // a unique name for the form
  form: 'login',
})(LoginForm);

export default LoginForm;
