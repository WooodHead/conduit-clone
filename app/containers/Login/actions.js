/*
 *
 * Login actions
 *
 */

import {
  DEFAULT_ACTION,
  ON_LOGIN_SUBMIT,
  AUTHENTICATION_REQUESTING,
  AUTHENTICATION_LOADED,
  AUTHENTICATION_LOADING_ERROR,
} from './constants';

export function defaultAction() {
  return {
    type: DEFAULT_ACTION,
  };
}

export function loginSubmit(email, password) {
  return {
    type: ON_LOGIN_SUBMIT,
    email,
    password,
  };
}

export function authenticationLoaded(user) {
  return {
    type: AUTHENTICATION_LOADED,
    user,
  };
}

export function authenticationLoadingError(err) {
  return {
    type: AUTHENTICATION_LOADING_ERROR,
    err,
  };
}

export function authenticationRequesting() {
  return {
    type: AUTHENTICATION_REQUESTING,
  };
}
