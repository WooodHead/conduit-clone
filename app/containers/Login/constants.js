/*
 *
 * Login constants
 *
 */

export const DEFAULT_ACTION = 'app/Login/DEFAULT_ACTION';
export const ON_LOGIN_SUBMIT = 'app/Login/ON_LOGIN_SUBMIT';
export const AUTHENTICATION_REQUESTING = 'app/Login/AUTHENTICATION_REQUESTING';
export const AUTHENTICATION_LOADED = 'app/Login/AUTHENTICATION_LOADED';
export const AUTHENTICATION_LOADING_ERROR = 'app/Login/AUTHENTICATION_LOADING_ERROR';
