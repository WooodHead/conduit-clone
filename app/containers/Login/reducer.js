/*
 *
 * Login reducer
 *
 */

import { fromJS } from 'immutable';
import {
  AUTHENTICATION_REQUESTING,
  AUTHENTICATION_LOADED,
  AUTHENTICATION_LOADING_ERROR,
} from './constants';

const initialState = fromJS({
  user: {
    fetching: false,
    error: false,
    data: fromJS({}),
  },
});

function loginReducer(state = initialState, action) {
  switch (action.type) {
    case AUTHENTICATION_LOADED:
      return state
        .setIn(['user', 'fetching'], false)
        .setIn(['user', 'data'], fromJS(action.user));
    case AUTHENTICATION_LOADING_ERROR:
      return state
        .setIn(['user', 'fetching'], false)
        .setIn(['user', 'error'], action.message);
    case AUTHENTICATION_REQUESTING:
      return state
        .setIn(['user', 'error'], false)
        .setIn(['user', 'fetching'], true);
    default:
      return state;
  }
}

export default loginReducer;
