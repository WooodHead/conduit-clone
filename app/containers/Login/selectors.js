import { createSelector } from 'reselect';

/**
 * Direct selector to the login state domain
 */
const selectLoginDomain = (state) => state.get('login');

/**
 * Other specific selectors
 */


/**
 * Default selector used by Login
 */

const makeSelectLoginUser = () => createSelector(
  selectLoginDomain,
  (substate) => substate.getIn(['user', 'data'])
);

const makeSelectLoginUserFetching = () => createSelector(
  selectLoginDomain,
  (substate) => substate.getIn(['user', 'fetching'])
);

const makeSelectLoginUserError = () => createSelector(
  selectLoginDomain,
  (substate) => substate.getIn(['user', 'error'])
);

export {
  selectLoginDomain,
  makeSelectLoginUser,
  makeSelectLoginUserFetching,
  makeSelectLoginUserError,
};
