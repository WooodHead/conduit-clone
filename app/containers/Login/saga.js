import { takeLatest, put, call } from 'redux-saga/effects';
import { ON_LOGIN_SUBMIT } from 'containers/Login/constants';
import {
  authenticationRequesting,
  authenticationLoaded,
  authenticationLoadingError,
} from 'containers/Login/actions';
import api from '../../api';

export function* onLoginSubmit(params) {
  yield put(authenticationRequesting());
  try {
    const payload = yield call(api.Auth.login, params);
    yield put(authenticationLoaded(payload.user));
  } catch (err) {
    yield put(authenticationLoadingError('incorrect email/password'));
  }
}

// Individual exports for testing
export default function* authentication() {
  yield takeLatest(ON_LOGIN_SUBMIT, onLoginSubmit);
}
