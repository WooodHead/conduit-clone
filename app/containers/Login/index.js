/**
 *
 * Login
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Helmet } from 'react-helmet';
import { FormattedMessage } from 'react-intl';
import { createStructuredSelector } from 'reselect';
import { compose } from 'redux';

import { Container } from 'reactstrap';

import injectSaga from 'utils/injectSaga';
import injectReducer from 'utils/injectReducer';
import makeSelectLogin from './selectors';
import reducer from './reducer';
import saga from './saga';
import messages from './messages';
import { loginSubmit } from './actions';
import LoginForm from './LoginForm';

export class Login extends React.PureComponent { // eslint-disable-line react/prefer-stateless-function
  constructor() {
    super();
    this.onsubmit = this.onsubmit.bind(this);
  }
  onsubmit(values) {
    this.props.onLoginSubmit(
      values.get('email'),
      values.get('password')
    );
  }
  render() {
    return (
      <Container>
        <Helmet>
          <title>Login</title>
          <meta name="description" content="Description of Login" />
        </Helmet>
        <FormattedMessage {...messages.header} />
        <LoginForm onSubmit={this.onsubmit} />
      </Container>
    );
  }
}

Login.propTypes = {
  onLoginSubmit: PropTypes.func,
};

const mapStateToProps = createStructuredSelector({
  login: makeSelectLogin(),
});

function mapDispatchToProps(dispatch) {
  return {
    onLoginSubmit: (email, password) => {
      dispatch(loginSubmit(email, password));
    },
  };
}

const withConnect = connect(mapStateToProps, mapDispatchToProps);

const withReducer = injectReducer({ key: 'login', reducer });
const withSaga = injectSaga({ key: 'login', saga });

export default compose(
  withReducer,
  withSaga,
  withConnect,
)(Login);
