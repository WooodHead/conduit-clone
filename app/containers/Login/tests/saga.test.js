/**
 * Test  sagas
 */

/* eslint-disable redux-saga/yield-effects */
import { takeLatest, put } from 'redux-saga/effects';
import { ON_LOGIN_SUBMIT } from 'containers/Login/constants';
import {
  authenticationLoadingError,
} from '../actions';
import authetication, { onLoginSubmit } from '../saga';

describe('defaultSaga Saga', () => {
  let onLoginSubmitGenerator;

  beforeEach(() => {
    onLoginSubmitGenerator = onLoginSubmit({
      email: 'email@web.com',
      password: 'sample password',
    });

    let putDescriptor = onLoginSubmitGenerator.next().value;
    expect(putDescriptor).toMatchSnapshot();

    const callDescriptor = onLoginSubmitGenerator.next({
      email: 'email@web.com',
      password: 'sample password',
    }).value;
    expect(callDescriptor).toMatchSnapshot();

    putDescriptor = onLoginSubmitGenerator.next({}).value;
    expect(putDescriptor).toMatchSnapshot();
  });

  it('should dispatch the authenticationLoadingError action if it requests throws an error', () => {
    const putDescriptor = onLoginSubmitGenerator.throw().value;
    expect(putDescriptor).toEqual(put(authenticationLoadingError('incorrect email/password')));
  });
});

describe('autheticationData Saga', () => {
  const autheticationData = authetication();
  it('should start task to watch for ON_LOGIN_SUBMIT action', () => {
    const takeLatestDescriptor = autheticationData.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(ON_LOGIN_SUBMIT, onLoginSubmit));
  });
});
