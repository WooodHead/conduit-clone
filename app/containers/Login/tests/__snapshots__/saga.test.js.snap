// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defaultSaga Saga should dispatch the authenticationLoadingError action if it requests throws an error 1`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "type": "app/Login/AUTHENTICATION_REQUESTING",
    },
    "channel": null,
  },
}
`;

exports[`defaultSaga Saga should dispatch the authenticationLoadingError action if it requests throws an error 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      Object {
        "email": "email@web.com",
        "password": "sample password",
      },
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`defaultSaga Saga should dispatch the authenticationLoadingError action if it requests throws an error 3`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "type": "app/Login/AUTHENTICATION_LOADED",
      "user": undefined,
    },
    "channel": null,
  },
}
`;
